### PR TITLE
Porting logcat to es6+ also adding functional and unit tests for it

### PR DESCRIPTION
**Pull request by moizjv**
- Logcat with es6+ and teen_process.
- Added android-22 and android-mnc
- @jlipps @sebv please review

### CODE REVIEW

File: lib/logcat.js [hunk 1]
```diff
@@ -1,118 +1,80 @@
...
-Logcat.prototype.onStderr = function (data) {
-  if (/execvp\(\)/.test(data)) {
-    logger.error('Logcat process failed to start');
-    if (!this.calledBack) {
-      this.calledBack = true;
-      this.onLogcatStart(new Error("Logcat process failed to start"));
-      return;
-    }
+  startCapture () {
+    return new Promise(async (resolve, reject) => {
```
Comment (sebv) [inline lib/logcat.js]: Async methods already return promises, no need to do that.

File: lib/logcat.js [hunk 2]
```diff
@@ -1,118 +1,80 @@
...
-    }
-  }.bind(this));
-  this.proc.on('exit', function (code, signal) {
-    logger.debug('Logcat terminated with code ' + code + ', signal ' + signal);
+class Logcat {
+  constructor (opts = {}) {
+    this.adb = opts.adb;
+    this.debug = opts.debug;
+    this.debugTrace = opts.debugTrace;
     this.proc = null;
```
Comment (sebv) [inline lib/logcat.js]: Most of `this.proc = null;` in the PR are unnecessary,

File: lib/logcat.js [hunk 3]
```diff
@@ -1,118 +1,80 @@
...
-    }
-  }
-  data = data.trim();
-  data = data.replace(/\r\n/g, "\n");
-  var logs = data.split("\n");
-  _.each(logs, function (log) {
-    log = log.trim();
-    if (log) {
-      var logObj = {
+  formatOuput (output, prefix) {
```
Comment (sebv) [inline lib/logcat.js]: This method do more than format output it may be worth doing some cleanup there.

File: test/unit/android-manifest-specs.js [hunk 4]
```diff
@@ -11,17 +11,17 @@ describe('android-manifest', async () => {
   let adb = new ADB();
   let withExecMock = (fn) => {
```
Comment (sebv) [inline test/unit/android-manifest-specs.js]: Same as last time, we could create a helper to do that in one line.

File: test/unit/logcat-specs.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import * as teen_process from 'teen_process';
+import events from 'events';
+import Logcat from '../../lib/logcat';
+
+chai.use(chaiAsPromised);
+
+describe('logcat', async () => {
+  let adb = {path: 'dummyPath', defaultArgs: []};
+  let logcat = new Logcat({adb: adb, debug: false, debugTrace: false});
+  let withMocks = (fn) => {
```
Comment (sebv) [inline test/unit/logcat-specs.js]: same

File: test/unit/logcat-specs.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    };
+  };
+  describe('startCapture', withMocks((mocks) => {
+    it('should correctly call subprocess and should resolve promise', async () => {
+      let conn = new events.EventEmitter();
+      conn.start = () => { };
+      mocks.teen_process.expects("SubProcess")
+        .once().withExactArgs('dummyPath', ['logcat'])
+        .returns(conn);
+      setTimeout(function () {
```
Comment (sebv) [inline test/unit/logcat-specs.js]: Can use nextTick

File: test/unit/logcat-specs.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+      logs.should.have.length.above(0);
+      mocks.teen_process.verify();
+    });
+    it('should correctly call subprocess and should reject promise', async () => {
+      let conn = new events.EventEmitter();
+      conn.start = () => { };
+      mocks.teen_process.expects("SubProcess")
+        .once().withExactArgs('dummyPath', ['logcat'])
+        .returns(conn);
+      setTimeout(function () {
```
Comment (sebv) [inline test/unit/logcat-specs.js]: Can do a promise returning function to do that and call it with  `produceLine().done();`, good promise exercise.

**Comment by sebv** [comment]
Few comments, lgtm.

File: test/unit/android-manifest-specs.js (hunk 4 above)
Comment (moizjv) [inline test/unit/android-manifest-specs.js]: Yeah I was thinking, how I can make it a utility with more flexibility.

File: test/unit/logcat-specs.js (hunk 6 above)
Comment (moizjv) [inline test/unit/logcat-specs.js]: I tried that, but some how does not work the same :(

File: lib/logcat.js [hunk 8]
```diff
@@ -1,118 +1,80 @@
...
-      this.onLogcatStart(new Error("Logcat process failed to start"));
-      return;
-    }
+  startCapture () {
+    return new Promise(async (resolve, reject) => {
+      log.debug("Starting logcat capture");
+      this.proc = new SubProcess(this.adb.path, this.adb.defaultArgs.concat(['logcat']));
+      await this.proc.start(0);
+      this.proc.on('exit', (code, signal) => {
+        log.info(`Logcat terminated with code ${code}, signal ${signal}`);
```
Comment (jlipps) [inline lib/logcat.js]: I would say this is worthy of a `log.error`

File: lib/logcat.js [hunk 9]
```diff
@@ -1,118 +1,80 @@
...
+        this.proc = null;
+      });
+      this.proc.on('lines-stderr', lines => {
+        for (let line of lines) {
+          if (/execvp\(\)/.test(line)) {
+            log.error('Logcat process failed to start');
+            reject(new Error(`Logcat process failed to start. stderr: ${line}`));
+          }
+          resolve();
```
Comment (jlipps) [inline lib/logcat.js]: why is this resolve here? is the one in `lines-stdout` not sufficient?

File: lib/logcat.js [hunk 10]
```diff
@@ -1,118 +1,80 @@
...
-  }
-  data = data.trim();
-  data = data.replace(/\r\n/g, "\n");
-  var logs = data.split("\n");
-  _.each(logs, function (log) {
-    log = log.trim();
-    if (log) {
-      var logObj = {
+  formatOuput (output, prefix) {
+    output.trim();
```
Comment (jlipps) [inline lib/logcat.js]: don't think this is doing anything. from MDN:

> The trim() method returns the string stripped of whitespace from both ends. trim() does not affect the value of the string itself.

File: lib/logcat.js [hunk 11]
```diff
@@ -1,118 +1,80 @@
-"use strict";
+import { SubProcess } from 'teen_process';
+import log from './logger';
```
Comment (jlipps) [inline lib/logcat.js]: do we want to do this? or maybe we just do:

```js
import { getLogger } from 'appium-logger';
const log = getLogger('Logcat');
```

that way we don't need to do `[LOGCAT STDERR]` below, it will just be blank for stdout and `[STDERR]` for stderr

**Comment by jlipps** [comment]
made a couple notes, overall looks really good!

File: test/unit/android-manifest-specs.js (hunk 4 above)
Comment (sebv) [inline test/unit/android-manifest-specs.js]: Call it like this, pass a map:

```js
withMock({adb, something}, fn(mocks) {
   ...
]);
```

File: lib/logcat.js (hunk 8 above)
Comment (moizjv) [inline lib/logcat.js]: When we stop it legitimately, it still triggers exit i think

File: lib/logcat.js (hunk 9 above)
Comment (moizjv) [inline lib/logcat.js]: either can be called first, so promise can be resolved at the earliest.

File: lib/logcat.js (hunk 10 above)
Comment (moizjv) [inline lib/logcat.js]: ohh yes it should have been `output = output.trim();`

File: lib/logcat.js (hunk 11 above)
Comment (moizjv) [inline lib/logcat.js]: hmm interesting. It is a choice that we have to make to have `ADB` as prefix or `LOGCAT` as prefix or as `ADB [LOGCAT ...` currently it is the third one.

Comment (jlipps) [inline lib/logcat.js]: right, I'm saying I think we could just have two prefixes, one for adb and one for logcat. from the user's perspective it doesn't matter that logcat is part of the adb package

File: lib/logcat.js (hunk 3 above)
Comment (moizjv) [inline lib/logcat.js]: do you mean i should change the name here? how about `outputHandler`?

File: lib/logcat.js (hunk 8 above)
Comment (jlipps) [inline lib/logcat.js]: then maybe before we stop it we should remove the exit listener? we're using SubProcess, so when we `await this.proc.stop();` we know that it has stopped after that, so we don't also need to be listening in that case. so what we could do is this to stop:

```js
this.proc.removeAllListeners('exit');
await this.proc.stop();
```

File: lib/logcat.js (hunk 9 above)
Comment (jlipps) [inline lib/logcat.js]: i guess what i don't like is that this is calling resolve() for every single line we ever get from stderr.

File: lib/logcat.js (hunk 1 above)
Comment (moizjv) [inline lib/logcat.js]: hmm what this intends to do is calling function successfully `awaits` till we resolve or reject the promise.

File: test/unit/android-manifest-specs.js (hunk 4 above)
Comment (moizjv) [inline test/unit/android-manifest-specs.js]: Ok, let me make all the changes in another pull request. :)
